### PR TITLE
Resimplify

### PR DESCRIPTION
**Pull request by CAD97**
cc @matklad; how does this API look now?

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,271 +1,353 @@
...
-///
-/// All this does is an `as`-cast between the erased slice type and a self pointer.
-#[macro_export]
-macro_rules! make_fat {
-    () => {
-        fn make_fat(ptr: *mut [$crate::priv_in_pub::Erased]) -> *mut Self {
-            ptr as *mut Self
-        }
-    };
+#[repr(C)]
```
Comment (matklad) [inline src/lib.rs:37]: API-wise, it would be cool if `#[repr(C)] ^ pub`, so that it's illegal for users to poke raw bytes. But that seems like overfoolpoofing.

File: src/lib.rs [hunk 2]
```diff
@@ -1,271 +1,353 @@
...
+    //     align(usize) < align(head) => head before length
+    //     align(head) < align(usize) => length before head
+    // We put length first because
+    //     a) it's much simpler to go from ErasedPtr to ptr-to-length
+    //     b) it's rare for types to have align > align(usize)
+    // For optimality, we should use repr(Rust) and pointer projection or offset_of!
+    // We don't do that for now since we can avoid the unsoundness of offset_of!,
+    // and offset_of! doesn't work for ?Sized types anyway.
+    // SAFETY: must be length of self.slice
+    length: usize,
```
Comment (matklad) [inline src/lib.rs]: len would be a more rustic name

File: src/lib.rs [hunk 3]
```diff
@@ -1,271 +1,353 @@
...
-    /// This materializes a `&`-reference on stable, thus unique mutation
-    /// after using this function is not necessarily allowed. Use
-    /// `make_fat_mut` instead if you have unique access.
-    unsafe fn make_fat_const(thin: ptr::NonNull<Erased>) -> ptr::NonNull<Self> {
-        let len = Self::get_length(thin.cast().as_ref());
-        ptr::NonNull::new_unchecked(Self::make_fat(make_slice(thin.as_ptr(), len) as *mut _))
+    unsafe fn fatten_const(ptr: ErasedPtr) -> ptr::NonNull<Self> {
+        let len = ptr::read(Self::length(ptr).as_ptr());
+        let slice = make_slice(ptr.as_ptr(), len);
+        ptr::NonNull::new_unchecked(slice as *const Self as *mut Self)
```
Comment (matklad) [inline src/lib.rs:67]: oh my, this works? So what happens here is that we have slice of somethings, and we cast it to a fat-ptr of our DST, and that reinterprets the pointer, while keeping metadata (len) in place? Cursed, but lovely!

File: src/lib.rs [hunk 4]
```diff
@@ -1,271 +1,353 @@
...
+        unsafe {
+            let ptr: ErasedPtr = ptr::NonNull::new(alloc(layout))
+                .unwrap_or_else(|| handle_alloc_error(layout))
+                .cast();
+            ptr::write(ThinData::<Head, SliceItem>::length(ptr).as_ptr(), len);
+            ThinData::fatten_mut(ptr.cast())
+        }
+    }
+
+    pub fn new(head: Head, slice: Vec<SliceItem>) -> Self {
```
Comment (matklad) [inline src/lib.rs]: I think servo's signature of `impl Itteraot<Item = SliceItem> + ExactSizeIterator` would be better here (we'll need to check that the iterator's lenght is not a lie, but that should be straighforwrd)

**Review by matklad — COMMENTED** [review]
LGMT, but I think we shouldn't hard-code `Vec` and use `ExactSizeIterator` instead

File: src/lib.rs (hunk 1 above)
Comment (CAD97) [reply src/lib.rs:37]: See also https://github.com/rust-lang/rust/issues/66401.

We have `length` first, so the offsets of `head`/`slice` aren't public anyway. This is not exposed well in rustdoc, but I think we can trust anyone dipping into `unsafe` for the `#[repr(C)]` of `ThinData` to notice that it's not actually public info.

File: src/lib.rs (hunk 3 above)
Comment (CAD97) [reply src/lib.rs:67]: Yep! Thank myrrlyn for clarifying what exactly is allowed here. This is specifically only sound because we're going from a slice-based DST to a slice-based DST, so the metadata is the same. That, plus our `[Erased(u8)]` is actually valid, because we have at least `len` bytes in the allocation. If you don't have that many bytes, you have to use `[()]` (but it still works if you do). (`ptr::slice_from_raw_parts[_mut]` may remove some of these requirements.)

File: src/lib.rs (hunk 4 above)
Comment (CAD97) [reply src/lib.rs]: `Itteraot`? 😆

There's two issues that mean I think just using `Vec` is fine.

1. `Box<[_]>` doesn't actually implement `IntoIterator`. If you want to into_iter a boxed slice, you need to `Vec::from` it (which is zero cost).
2. By using `Vec` here, we get to use `ptr::copy_nonoverlapping` rather than iterating the source slice. This makes the code not have to worry about panics in the iterator and also guarantees that it's just a single copy rather than piecewise.
3. I'd personally use `TrustedLen` for `new` if it took `ExactSizeIterator`. Perhaps we could offer a secondary constructor that takes an `ExactSizeIterator + ?TrustedLen` if we find it's desired?

Can you think of any use cases in rust-analyzer/rowan that would actually use `ThinArc::new` with a slice that isn't `Vec<_>` or `Box<[_]>`?

All of the places that currently `collect::<Box[_]>` in ra go through `fn with_children(&SyntaxNode, Box<[GreenElement]>)`, and that requires iterating through `new_children` to get the text length as well as passing to `GreenNode::new`.

The "best" signature that `with_children` could take that accepts all of its callers without `collect` would be

```
fn with_children<I>(parent: &SyntaxNode, new_children: I)
where
    I: Iterator<Item=GreenElement> + ExactSizeIterator + TrustedLen,
    I: Clone,
```

and that's only if `fn insert_children` goes from taking `to_insert: &mut dyn Iterator<Item=SyntaxElement>` to `to_insert: &mut dyn CloneableIterator<Item=SyntaxElement>` (where `trait CloneableIterator: Iterator + Clone; impl CloneableIterator for impl Iterator + Clone;`).

TL;DR I don't think it's worth taking something other than `Vec` here.

File: src/lib.rs (hunk 3 above)
Comment (CAD97) [reply src/lib.rs:67]: Actually, ZSTs were mishandled (fix: f53f221d1dcbca0fe47f3a11587efbed3d7420df) by this. (Oh, the pain of making `Erased` and falsely thinking it'd just magically be size 0 and stride 1 depending on which I wanted.)

File: src/lib.rs (hunk 4 above)
Comment (matklad) [reply src/lib.rs]: >Can you think of any use cases in rust-analyzer/rowan that would actually use ThinArc::new with a slice that isn't Vec<_> or Box<[_]>?

See https://github.com/rust-analyzer/rowan/pull/36. Basically, green builder already allocates all the nodes in a single vec, and it would be cool if we can drain that vec into a green node without extra allocation. 

Hm, oh, I guess we still have to do one allocation because of an intermediate Box =/

Hm, ok, I think a `Vec` is fine then, though `ExactSizeIterator` does look like a canonical signature here///

Comment (CAD97) [reply src/lib.rs]: I've switched it over to take `IntoIterator`. You've won me over and it nicely supports that use case, and is forward-good with directly allocating `Arc`s.

I'm tempted to add a `--cfg abandon-morals` that does it anyway 😛
